### ISSUE TITLE
[Bug]: PC0037 is firing if the table type is temporary

### ISSUE THREAD

**Issue by nickgoddard777** [ALCops/Analyzers#422]
### Analyzer

PlatformCop

### Issue type

False positive (diagnostic fires when it shouldn't)

### Rule ID

PC0037

### Description

I have a table that is TableType temporary. When assigning values to fields in this table rule PC0037 is firing.

<img width="729" height="345" alt="Image" src="https://github.com/user-attachments/assets/c4a71630-2bc8-4efa-af6c-857e480d7400" />

<img width="901" height="193" alt="Image" src="https://github.com/user-attachments/assets/3877c3ac-152e-46b1-a1cb-d0fb4791e5ad" />

### AL code to reproduce

```al
table 50500 "Finance Transaction"
{
    TableType = Temporary;
    fields
    {
        field(1; "Transaction Id"; Code[50])
        {
            Caption = 'transactionId';
            ToolTip = 'Specifies the Amazon transaction Id.';
            NotBlank = true;
        }
    }
}

codeunit 50500 "Transaction Functions"
{
    procedure SetValue(Value: Code[20])
    begin
        Rec."Transaction Id" := Value;
    end;
}
```

**Comment by MODUSCarstenScholling** [comment]
A similar issue has already been addressed. @Arthurvdv, when you have a moment, could you please share the reasoning behind also applying this to TableType = Temporary in [this PR](https://github.com/ALCops/Analyzers/issues/369#issuecomment-4841847108)?

My understanding is that Microsoft now treats temporary tables and tables declared with TableType = Temporary equivalently.

**Comment by Arthurvdv** [comment]
Thanks @nickgoddard777 for the report and @MODUSCarstenScholling for the pointer to #369. I went back to the SDK source, the Base Application (w1-28) and the Microsoft docs before answering, because the reasoning in #369 deserved a proper substantiation. Short version: **I don't consider this a false positive, and I'd like to make the rule *more* consistent here rather than looser** — but I'm laying out the evidence so you can push back.

### Three cases, and what PC0037 does today

| | Declaration | Today |
|---|---|---|
| A | `Record X temporary` where X is a normal table | silent |
| B | `Record X` where X has `TableType = Temporary` | **fires** (this issue) |
| C | `Record X temporary` where X has `TableType = Temporary` | silent |

### Why B stays in scope

**1. The design intent argument (discussion #259).** Christian Hovenbitzer's point was: exclude temporary *variables* (case A — a real table used as a scratch buffer, where running `OnValidate` could have side effects you didn't sign up for), but keep *inherently* temporary tables in scope, because whoever wrote an `OnValidate` on a `TableType = Temporary` table knew it was temporary. That logic is intended to run. Bypassing it is exactly what PC0037 exists to flag.

**2. Base Application evidence backs that up.** Of the 211 `TableType = Temporary` tables in w1-28, 22 have field `OnValidate` triggers (75 in total). Roughly 30 of those 75 do real work beyond the record itself: `Employee Time Reg Buffer."Employee Id"` calls `Employee.Modify()` and creates a `Resource`; `Email Attachments.AttachmentDefaultStatus` modifies the persisted `Email Scenario Attachments`; `Price Asset."Unit of Measure Code"` raises an integration event and `TestField`s; `VAT Amount Line."VAT Amount"` calls into `NonDeductibleVAT`. A direct `:=` on any of those silently skips business logic — the same failure mode as on a normal table.

**3. Extensibility.** `TableType = Temporary` tables are extensible (the compiler's `SemanticFacts.IsTableTypeExtensible` includes them), so a tableextension from another app can add `OnBeforeValidate`/`OnAfterValidate` to your field. That's the "all tables / all fields / always" argument from the TechDays session, and it applies unchanged.

**4. Consistency inside ALCops.** PC0027 (*don't call `Validate()` on temporary records*) already **exempts** `TableType = Temporary` tables — ALCops's position is that validation on an inherently-temporary table is legitimate. Dropping B from PC0037 would make the two rules contradict each other.

### On "Microsoft treats them equivalently"

At *runtime*, yes — the docs say `TableType = Temporary` "has the same effect as using a temporary record variable". But the compiler does not: `IRecordTypeSymbol.Temporary` is set only from the `temporary` keyword (`Binder`/`RecordTypeSymbol`), the two forms are distinct record types, and Microsoft's own CodeCop rules have to add the `TableType` check by hand where it matters (AA0073/AA0237 require the `Temp` prefix on a `Record X` variable when X is `TableType = Temporary`, keyword or not). So "equivalent" is a runtime fact, not a compiler or design rule — and the design question is precisely what PC0037 is about.

### The inconsistency I *do* want to fix: case C

Right now the rule fires on B but not on C, i.e. on the same `TableType = Temporary` table the outcome depends on whether you typed a redundant `temporary` keyword. In the Base Application that keyword is sprinkled by history, not intent (`VAT Amount Line` has it on 37% of declarations, `Price Asset` on 10%, `Price Line Filters` on 0%). That's a loophole, not a feature. Proposal: **when the base table is `TableType = Temporary`, decide by the table type and ignore the variable keyword → C fires like B.** Case A stays excluded as agreed in #259.

### What I'm not proposing

- Excluding B because 90% of temp tables have no triggers — the same is true for plenty of normal tables and the rule deliberately doesn't gate on trigger presence (extensions).
- Excluding B because `Validate()` might have unwanted side effects — that's the case-A argument; on an inherently temporary table the side effects are the designed behaviour.

For trigger-less buffer tables where the warning genuinely adds nothing, `#pragma warning disable PC0037` on the block (or a ruleset entry) remains the intended escape hatch. I'll also fix the docs page, which today only mentions the `temporary` variable exception and never says `TableType = Temporary` is in scope — that omission is on me.

If you have a scenario where running `OnValidate` on an inherently-temporary table is *wrong by design* (not just unnecessary), I'd genuinely like to see it before I implement the case-C change.

*Created with Claude Fable 5*

**Comment by Arthurvdv** [comment]
I’ve let Claude take a look at this, where I agree with the finding of it and I don't think this is a false positive.

Why shouldn’t we apply `.Validate()` when the table type is Temporary? Having `TableType = Temporary` on the object is by design, so I would expect any `OnValidate()` trigger to work as intended in that context as well.

**Comment by Arthurvdv** [comment]
@nickgoddard777 and @MODUSCarstenScholling did you happen to have a chance to take a look at my feedback?

**Comment by nickgoddard777** [comment]
@Arthurvdv Yes, sorry I forgot to respond. First thanks for taking the time to come up with such a detailed update. I think after reading the full explanation, what you've said makes sense. In hindsight I think for the times where we don't want to use `.Validate()` we can add  `#pragma warning disable PC0037`

**Comment by MODUSCarstenScholling** [comment]
@Arthurvdv Unfortunately, it is not quite that simple. With issues taking months to be marked as resolved and overlapping issues on the same topic the current position of Microsoft is rather difficult to determine.

This was first raised in microsoft/AL#7811 and is also mentioned in the AL [changelog](https://marketplace.visualstudio.com/items/ms-dynamics-smb.al/changelog). The newer issue, microsoft/AL#8221, has been accepted, but its exact scope remains unclear: does it apply only to local variables, or to parameters as well?

Personally, I do not think TableType = Temporary should affect the naming rule. A table designed as temporary should not require the Temp prefix; only variables based on regular tables and declared as temporary should.

But perhaps Microsoft still has a surprise in store for us.

**But we know: If a table is designed to be temporary, it is reasonable to assume that calling `Validate()` will not cause any unintended side effects. Therefore, `TableType = Temporary` tables should remain within the scope of PC0037, regardless of whether the record variable also uses the redundant `temporary` keyword.**

**Comment by Arthurvdv** [comment]
@MODUSCarstenScholling Fair point on the AA0237 / `Temp` prefix naming discussion. Depending on the outcome of the issue you've linked I'm even considering if we need an rule for B where the `temporary` is omitted on the variable of a Record has `TableType = Temporary` set, but I think that's a separate topic from what PC0037 is about.

On the PC0037 side: it sounds like we're all in agreement to accept the proposal; `TableType = Temporary` tables stay in scope for PC0037, and case C (redundant `temporary` keyword on a `TableType = Temporary` table) should fire as well. I'll go ahead and merge the PR on this.

**Comment by Arthurvdv** [comment]
This should now be resolved with the release of `v1.2.0` of ALCops.

When you have a chance, could you update and verify whether the issue is fixed?

If everything is working as expected, feel free to close this issue. If you still run into the problem, please leave the issue open and share any additional context, code samples, or reproduction steps that might help me investigate further.
